### PR TITLE
[CALCITE] RANK function call with sorting expression parameters

### PR DESCRIPTION
**Pull request by michaelvdow**

### CODE REVIEW

File: babel/src/main/codegen/includes/parserImpls.ftl [hunk 1]
```diff
@@ -987,3 +987,51 @@ SqlNode AlternativeTypeConversionQuery(SqlNode query) :
...
+// Parse RANK function call with sorting expressions.
+SqlNode RankFunctionCallWithParams() :
+{
+    final SqlFunctionCategory funcType = SqlFunctionCategory.USER_DEFINED_FUNCTION;
+    final SqlNodeList orderByList = new SqlNodeList(SqlParserPos.ZERO);
+    final SqlIdentifier qualifiedName;
+    final SqlNode rankCall;
+    final SqlNode over;
+    SqlNode e;
+    Span s;
```
Comment (ArthurInTheShell) [inline babel/src/main/codegen/includes/parserImpls.ftl:1016]: Would it be reasonable to have the variable "s" and "s1" to be final?

Comment (michaelvdow) [reply babel/src/main/codegen/includes/parserImpls.ftl:1016]: Yeah those should probably be final if they're not being reassigned

File: babel/src/main/codegen/includes/parserImpls.ftl [hunk 2]
```diff
@@ -987,3 +987,51 @@ SqlNode AlternativeTypeConversionQuery(SqlNode query) :
...
+    <LPAREN> { s1 = span(); }
+    e = RankSortingExpression() { orderByList.add(e); }
+    (
+        <COMMA> e = RankSortingExpression() {
+            orderByList.add(e);
+        }
+    )*
+    <RPAREN> {
+        over = SqlWindow.create(null, null, SqlNodeList.EMPTY, orderByList,
+                SqlLiteral.createBoolean(false, SqlParserPos.ZERO), null, null, null, s1.end(this));
```
Comment (dasch-1) [inline babel/src/main/codegen/includes/parserImpls.ftl:1034]: nit: This indentation looks a bit off.

File: babel/src/main/codegen/includes/parserImpls.ftl [hunk 3]
```diff
@@ -987,3 +987,51 @@ SqlNode AlternativeTypeConversionQuery(SqlNode query) :
         return SqlStdOperatorTable.CAST.createCall(s.end(this), args);
     }
 }
+
+SqlNode RankSortingExpression() :
```
Comment (dasch-1) [inline babel/src/main/codegen/includes/parserImpls.ftl:991]: Suggest adding a comment stating that the "RANK(foo)" function has a default ordering of DESC and the RANK() OVER ..." clause has a default ordering of ASC.

File: core/src/main/codegen/templates/Parser.jj [hunk 4]
```diff
@@ -5666,6 +5666,11 @@ SqlNode BuiltinFunctionCall() :
         node = ${method} { return node; }
     |
 </#list>
+<#if parser.includeRankWithSortingExpressions>
+        LOOKAHEAD(<RANK>, { getToken(3).kind != RPAREN })
```
Comment (dasch-1) [inline core/src/main/codegen/templates/Parser.jj:5670]: Nice.

File: babel/src/main/codegen/includes/parserImpls.ftl (hunk 2 above)
Comment (michaelvdow) [reply babel/src/main/codegen/includes/parserImpls.ftl]: Yeah it was, updated now

File: babel/src/main/codegen/includes/parserImpls.ftl (hunk 3 above)
Comment (michaelvdow) [reply babel/src/main/codegen/includes/parserImpls.ftl:995]: Comment added

File: babel/src/main/codegen/includes/parserImpls.ftl [hunk 5]
```diff
@@ -988,6 +988,10 @@ SqlNode AlternativeTypeConversionQuery(SqlNode query) :
     }
 }
 
+/**
+ * The RANK function call with sorting expressions has a default ordering of DESC
+ * while the RANK() OVER function call has a default ordering of ASC
```
Comment (dasch-1) [inline babel/src/main/codegen/includes/parserImpls.ftl:993]: nit: missing punctuation

Comment (michaelvdow) [reply babel/src/main/codegen/includes/parserImpls.ftl]: Fixed

**Pull request merged**